File: problems/3593-minimum-increments-to-equalize-leaf-paths/solution.java
import java.util.*;

class Solution {
    public int minIncrease(int n, int[][] edges, int[] cost) {
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
        }
        for (int[] edge : edges) {
            adj.get(edge[0]).add(edge[1]);
            adj.get(edge[1]).add(edge[0]);
        }

        List<Integer> leafPathSums = new ArrayList<>();
        dfs(0, adj, cost, 0, new boolean[n], leafPathSums);

        int maxPathSum = 0;
        for (int sum : leafPathSums) {
            maxPathSum = Math.max(maxPathSum, sum);
        }

        int increases = 0;
        for (int sum : leafPathSums) {
            increases += (maxPathSum - sum);
        }

        return increases;
    }

    private void dfs(int node, List<List<Integer>> adj, int[] cost, int currentSum, boolean[] visited, List<Integer> leafPathSums) {
        visited[node] = true;
        currentSum += cost[node];

        boolean isLeaf = true;
        for (int neighbor : adj.get(node)) {
            if (!visited[neighbor]) {
                isLeaf = false;
                dfs(neighbor, adj, cost, currentSum, visited.clone(), leafPathSums);
            }
        }

        if (isLeaf) {
            leafPathSums.add(currentSum);
        }
    }
}

File: problems/3593-minimum-increments-to-equalize-leaf-paths/solution.py
class Solution:
    def minIncrease(self, n: int, edges: List[List[int]], cost: List[int]) -> int:
        graph = [[] for _ in range(n)]
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)

        leaves = []

        def dfs(node, parent, path_sum):
            is_leaf = True
            for neighbor in graph[node]:
                if neighbor != parent:
                    is_leaf = False
                    dfs(neighbor, node, path_sum + cost[neighbor])
            if is_leaf:
                leaves.append(path_sum)

        dfs(0, -1, cost[0])

        if not leaves:
            return 0

        max_path_sum = max(leaves)
        ans = 0
        for path_sum in leaves:
            ans += max_path_sum - path_sum
        
        increases = {}
        
        def dfs2(node, parent, path_sum, increase_count):
            is_leaf = True
            for neighbor in graph[node]:
                if neighbor != parent:
                    is_leaf = False
                    dfs2(neighbor, node, path_sum + cost[neighbor], increase_count)
            if is_leaf:
                return
            
            
        
        return ans // (max_path_sum - min(leaves)) if max_path_sum != min(leaves) else 0